test/puppetlabs/puppetdb/jdbc_test.clj

```diff
@@ -1,36 +1,46 @@
 (ns puppetlabs.puppetdb.jdbc-test
   (:require [puppetlabs.puppetdb.jdbc :as subject]
-            [clojure.java.jdbc.deprecated :as sql]
             [puppetlabs.puppetdb.fixtures :as fixt]
             [clojure.test :refer :all]
+            [puppetlabs.puppetdb.jdbc :as jdbc]
             [puppetlabs.puppetdb.testutils :refer [test-db]]
             [puppetlabs.puppetdb.testutils.db :refer [antonym-data with-antonym-test-database insert-map *db-spec*]]))
 
 (use-fixtures :each with-antonym-test-database)
 
+(defn- full-sql-exception-msg [ex]
+  (apply str (take-while identity (iterate #(.getNextException %) ex))))
+
 (deftest pool-construction
 
   (testing "can construct pool with numeric usernames and passwords"
-    (let [pool (-> (test-db)
+    (let [pool (-> *db-spec*
                    (assoc :username "1234" :password "1234")
                    fixt/defaulted-write-db-config
                    subject/pooled-datasource)]
       (.close (:datasource pool))))
 
   (testing "writes not allowed on read-only pools"
-    (let [write-pool (subject/pooled-datasource (fixt/defaulted-write-db-config *db-spec*))
-          read-pool (subject/pooled-datasource (fixt/defaulted-read-db-config (assoc *db-spec* :read-only? true)))]
-
-      (subject/with-transacted-connection write-pool
-        (insert-map {"foo" 1})
-        (is (= [{:value "1"}] (subject/query-to-vec "SELECT value FROM test WHERE key='foo'"))))
-
-      (subject/with-transacted-connection read-pool
-        (is (thrown-with-msg? java.sql.SQLException #"read-only.*transaction"
-                              (insert-map {"bar" 1}))))
-
-      (.close (:datasource write-pool))
-      (.close (:datasource read-pool)))))
+    (let [write-pool (-> *db-spec*
+                         fixt/defaulted-write-db-config
+                         subject/pooled-datasource)]
+      (with-open [_ (:datasource write-pool)]
+        (subject/with-transacted-connection write-pool
+          (insert-map {"foo" 1})
+          (is (= [{:value "1"}]
+                 (subject/query-to-vec
+                  "SELECT value FROM test WHERE key='foo'"))))))
+    (let [read-pool (-> (assoc *db-spec* :read-only? true)
+                        fixt/defaulted-read-db-config
+                        subject/pooled-datasource)]
+      (with-open [_ (:datasource read-pool)]
+        (subject/with-transacted-connection read-pool
+          (let [msg (try
+                      (insert-map {"bar" 1})
+                      ""
```

Comment:
One change here was to use with-open so that we always close.  Another was to separate\ the read and write pool tests, since they didn't share the connections, and finally, the "read-only" string moved.  It's no longer in the first exception in the "get cause" list, so I added the full-sql-exception-msg function to flatten the text (which might be what you were really asking about...).